File: week 7/04-04-2026 weekend/multi catalog bill management/BillGenerator.Domain/BillingMath.cs
using BillGenerator.Domain.Entities;
using BillGenerator.Domain.Enums;

namespace BillGenerator.Domain;

/// <summary>
/// Single place for subtotal, discount, tax, and grand total so API and UI stay aligned.
/// Rule: tax applies to (subtotal − discount), then grand total = taxable base + tax.
/// </summary>
public static class BillingMath
{
    public static decimal LineTotal(int quantity, decimal unitPrice) =>
        Math.Round(quantity * unitPrice, 2, MidpointRounding.AwayFromZero);

    public static decimal Subtotal(IEnumerable<BillLine> lines) =>
        Math.Round(lines.Sum(l => l.LineTotal), 2, MidpointRounding.AwayFromZero);

    public static decimal DiscountAmount(decimal subtotal, DiscountType type, decimal value)
    {
        if (type == DiscountType.None || subtotal <= 0)
            return 0;

        if (type == DiscountType.Percent)
        {
            var pct = Math.Clamp(value, 0, 100);
            return Math.Round(subtotal * (pct / 100m), 2, MidpointRounding.AwayFromZero);
        }

        // Fixed: cannot discount more than subtotal
        return Math.Round(Math.Min(value, subtotal), 2, MidpointRounding.AwayFromZero);
    }

    public static decimal TaxableBase(decimal subtotal, decimal discountAmount) =>
        Math.Max(0, Math.Round(subtotal - discountAmount, 2, MidpointRounding.AwayFromZero));

    public static decimal TaxAmount(decimal taxableBase, decimal taxRatePercent)
    {
        if (taxableBase <= 0 || taxRatePercent <= 0)
            return 0;
        return Math.Round(taxableBase * (taxRatePercent / 100m), 2, MidpointRounding.AwayFromZero);
    }

    public static decimal GrandTotal(decimal taxableBase, decimal taxAmount) =>
        Math.Round(taxableBase + taxAmount, 2, MidpointRounding.AwayFromZero);

    /// <summary>Recomputes each line's LineTotal from quantity × unit price.</summary>
    public static void RefreshLineTotals(IEnumerable<BillLine> lines)
    {
        foreach (var line in lines)
            line.LineTotal = LineTotal(line.Quantity, line.UnitPrice);
    }
}

File: week 7/04-04-2026 weekend/multi catalog bill management/bill-generator-ui/src/utils/billingMath.js
/**
 * Client-side totals (must match server `BillingMath` in the API).
 * Order: line totals → subtotal → discount → tax on (subtotal − discount) → grand total.
 */

function round2(n) {
  return Math.round((Number(n) + Number.EPSILON) * 100) / 100
}

export function lineTotal(quantity, unitPrice) {
  return round2(Number(quantity) * Number(unitPrice))
}

export function subtotal(lines) {
  return round2(lines.reduce((s, l) => s + Number(l.lineTotal ?? 0), 0))
}

export function discountAmount(sub, discountType, discountValue) {
  if (discountType === 0 || sub <= 0) return 0
  if (discountType === 1) {
    const pct = Math.min(100, Math.max(0, Number(discountValue)))
    return round2(sub * (pct / 100))
  }
  return round2(Math.min(Number(discountValue), sub))
}

export function taxableBase(sub, discountAmt) {
  return round2(Math.max(0, sub - discountAmt))
}

export function taxAmount(taxable, taxRatePercent) {
  if (taxable <= 0 || !taxRatePercent) return 0
  return round2(taxable * (Number(taxRatePercent) / 100))
}

export function grandTotal(taxable, taxAmt) {
  return round2(taxable + taxAmt)
}

/** @param {object} bill — detail DTO with lines, discountType, discountValue, taxRatePercent */
export function computeTotals(bill) {
  const lines = bill.lines ?? []
  const sub = subtotal(lines)
  const disc = discountAmount(sub, bill.discountType, bill.discountValue)
  const base = taxableBase(sub, disc)
  const tax = taxAmount(base, bill.taxRatePercent)
  const grand = grandTotal(base, tax)
  return {
    subtotal: sub,
    discountAmount: disc,
    taxableBase: base,
    taxAmount: tax,
    grandTotal: grand,
  }
}
